### DISCUSSION TITLE
Adding a few functions to the python bindings

### DISCUSSION THREAD

**Discussion by saeedhd96** [mitsuba-renderer/mitsuba3#416] [Feature requests]
Hi,

I can (and I normally do) add new python bindings for the functions in C++ that are not already introduced in the bindings.
However, there are some functions that I think are very useful for everybody and would be awesome if you could add it too, so everybode doesn't have to touch the mitsuba code.

1.  In ShapePtr::sample_position()
2. In BSDF::eval_diffuse_reflectance()

Thanks,
-Saeed

**Comment by saeedhd96** [comment]
Oh I just noticed the eval_diffuse_reflectance is recently added. So just number 1.

**Comment by saeedhd96** [comment]
This is how I do it myself in the shape_v.cpp:

https://github.com/mitsuba-renderer/mitsuba3/blob/199b607d6e878d943cf908443bca5b0c399d18bd/src/render/python/shape_v.cpp#L135



    if constexpr (ek::is_array_v<ShapePtr>) {
        py::object ek       = py::module_::import("enoki"),
                   ek_array = ek.attr("ArrayBase");

        py::class_<ShapePtr> cls(m, "ShapePtr", ek_array);
        //BEGIN CHANGED
        cls
        .def("sample_position",
                [](ShapePtr ptr, Float time, const Point2f &sample, Mask active) {
                    return ptr->sample_position(time, sample, active);
                },
                "time"_a, "sample"_a, "active"_a = true,
                D(Shape, sample_position));        
        //END CHANGED        
        bind_shape_generic<ShapePtr>(cls);
    }


But It'd be nice to have this in the version of mitsuba we get through pip too (pip is more convienient since it can be easily imported on any cluster and container)

**Comment by saeedhd96** [comment]
Hi, just a quick reminder on this. Is this possible to do?

This is helpful when calling sample position on the result of `scene.shapes_dr()`,  'cause curretnly it throws an error saying `ShapePtr` doesn't have `sample_position()`.

**Comment by njroussel** [comment]
Indeed! Nice catch, I'll add that right away.

**Reply by njroussel** [reply]
The missing bindings have been added in this commit: bdce9509f0504163678e81c6afdd7a8bc9c45340
They will be available in the pip build of the next release.

**Reply by saeedhd96** [reply]
Amazing! Thanks!
